File: 0297/Codec.php
<?php
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($value) { $this->val = $value; }
 * }
 */

class Codec {
    function __construct() {

    }

    /**
     * @param TreeNode $root
     * @return String
     */
    function serialize($root) {
        $queue = [$root];
        $arr = [];

        while (!empty($queue)) {
            $node = array_shift($queue);

            if ($node) {
                $arr[] = $node->val;
                $queue[] = $node->left;
                $queue[] = $node->right;
            } else {
                $arr[] = null;
            }
        }

        return json_encode($arr);
    }

    /**
     * @param String $data
     * @return TreeNode
     */
    function deserialize($data) {
        $arr = json_decode($data);
        $root = new TreeNode($arr[0]);
        $queue = [$root];

        $i = 1;
        while (!empty($queue)) {
            $node = array_shift($queue);

            if ($arr[$i] !== null) {
                $node->left = new TreeNode($arr[$i]);
                $queue[] = $node->left;
            }
            $i++;

            if ($arr[$i] !== null) {
                $node->right = new TreeNode($arr[$i]);
                $queue[] = $node->right;
            }
            $i++;
        }

        return $root;
    }
}

/**
 * Your Codec object will be instantiated and called as such:
 * $obj = Codec();
 * $data = $obj->serialize($root);
 * $ans = $obj->deserialize($data);
 */

File: 0297/Codec.go
package _0297

import (
    "fmt"
    "github.com/raw34/leetcode/runtime"
    "strconv"
    "strings"
)

type Codec struct {
}

func Constructor() Codec {
    return Codec{}
}

// Serializes a tree to a single string.
func (this *Codec) serialize(root *runtime.TreeNode) string {
    res := ""
    var dfs func(root *runtime.TreeNode)
    dfs = func(root *runtime.TreeNode) {
        if root == nil {
            res += "null,"
            return
        }
        res += fmt.Sprintf("%d,", root.Val)
        dfs(root.Left)
        dfs(root.Right)
    }
    dfs(root)

    return res
}

// Deserializes your encoded data to tree.
func (this *Codec) deserialize(data string) *runtime.TreeNode {
    values := strings.Split(data, ",")
    var build func() *runtime.TreeNode
    build = func() *runtime.TreeNode {
        str := values[0]
        values = values[1:]
        if str == "null" {
            return nil
        }
        val, _ := strconv.Atoi(str)
        root := &runtime.TreeNode{Val: val}
        root.Left = build()
        root.Right = build()

        return root
    }

    return build()
}
